### PR TITLE
NodeGuy/709 improve automate deployment

### LINKED ISSUE #709: Automate deployment.
Every push to `master` should automatically deploy a new release to GitHub.

This can be done on CircleCI.
It would be good to run it by Jessy before implementing.

### PR DESCRIPTION
**Pull request by faboweb**
Closes #709

❤️ Thank you!

### CODE REVIEW

**Comment by jolesbi** [comment]
something we keep missing during releases is updating the version "About Cosmos Voyager" section - can this be automated?

screen shot 2018-05-25 at 2 32 32 pm

**Comment by NodeGuy** [comment]
> something we keep missing during releases is updating the version "About Cosmos Voyager" section - can this be automated?

That should be picking the version number out of `package.json`.  @nylira Where's the code for the `About` window?  I'm having trouble finding it.

File: ci/release.js [hunk 1]
```diff
@@ -138,4 +136,12 @@ async function main() {
   )
 }
 
-main().then(null, console.error)
+if (require.main === module) {
```
Comment (faboweb) [inline ci/release.js]: nice way to enable testing

File: ci/release.js [hunk 2]
```diff
@@ -28,6 +28,19 @@ function updateChangeLog(changeLog, newVersion, now) {
 const updatePackageJson = (packageJson, version) =>
   Object.assign({}, packageJson, { version })
 
+const pushCommit = (token, branch, tag) =>
```
Comment (faboweb) [inline ci/release.js]: 👍

File: ci/release.js [hunk 3]
```diff
@@ -95,30 +95,22 @@ async function main() {
   )
 
   console.log("--- Committing release changes ---")
-
-  const VoyagerCommit = execSync("git rev-parse --short=6 HEAD")
-    .toString()
-    .trim()
-
-  console.log("Voyager commit:", VoyagerCommit)
```
Comment (faboweb) [inline ci/release.js]: The idea behind fetching and posting the commit was to enable reproducability of the build. `HEAD` will change over time. Do you think this is not needed? Is there another way to post the commit?

**Comment by faboweb** [comment]
:/ Script currently gets triggered on every commit. See all the "Bump version for release."

File: ci/release.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    __dirname + "/../package.json",
+    JSON.stringify(newPackageJson, null, 2),
+    "utf8"
+  )
+
+  console.log("--- Committing release changes ---")
+  pushCommit(process.env.GIT_BOT_TOKEN, newVersion)
+  console.log("SDK commit:", releaseConfig.SDK_COMMIT) // TODO put in config.toml?
+
+  if (!process.env.SKIP_BUILD) {
```
Comment (NodeGuy) [inline ci/release.js]: It doesn't make sense to me to skip the building if we're bumping the version number.  If we don't build, the app will have the old version number.  Is that your intention?

Comment (faboweb) [reply ci/release.js]: Ah sorry, I left that in. I used the flag to quickly test the build script, skipping the long build step.

**Comment by faboweb** [comment]
Further improvement potential: https://github.com/cosmos/voyager/issues/760

**Comment by faboweb** [comment]
Where are we on this? Which steps are missing? Where can I help?

**Comment by NodeGuy** [comment]
All steps are in place.  It's [currently failing on CircleCI](https://circleci.com/gh/cosmos/voyager/2884#config/containers/0) with this error:

```
Build-agent version 0.0.5706-58f0e71 (2018-05-14T22:35:09+0000)
Configuration errors: 1 error occurred:

* Cannot find a job named `build` to run in the `jobs:` section of your configuration file.
If you expected a workflow to run, check your config contains a top-level key called 'workflows:'
```

There is a top-level key called 'workflows:' so I'm trying to figure out why it's not working.  It may be [a bug in CircleCI](https://discuss.circleci.com/t/if-you-expected-a-workflow-to-run-check-your-config-contains-a-top-level-key-called-workflows/16798/13).

I don't see any place you can help until I'm past this hurdle; thank you for asking.

**Comment by nylira** [comment]
I tried to debug this but also got stuck on that same configuration error.

**Comment by NodeGuy** [comment]
Thanks, I labelled my last commit `WIP` because it's still in progress and I don't want anyone else to add changes while it's in an unstable state.  I'm reverting your merge.

**Comment by NodeGuy** [comment]
It no longer makes sense to trigger this from a push to `master` now that you're making a new commit and pushing to `master` again because that just creates an endless loop.

I'm fine with the script creating a new commit and pushing to `master` but that means we should figure out a different way to trigger a release.  Unfortunately CircleCI doesn't have a manual trigger.  They do have scheduled triggers and manual approval of jobs—maybe it would be fruitful to combine the two.  Or we could push to a different branch to trigger a release, like `make-release` or something.

**Comment by faboweb** [comment]
Another idea would is checking if it is a merge/commit by our bot and do logic according to that. Pushing to `make-release` also sounds like a valid option.

**Comment by NodeGuy** [comment]
OK, I'm going with pushing to `release` to trigger a release to `master`.

It's now failing to push to `master`.  How do you want to handle this?

```
--- Committing release changes ---
remote: error: GH006: Protected branch update failed for refs/heads/master.        
remote: error: 3 of 3 required status checks are expected.        
To https://6a3d469f30df9b9597fdcd91442ab9f973a9dbde@github.com/cosmos/voyager.git
 ! [remote rejected] HEAD -> master (protected branch hook declined)
error: failed to push some refs to 'https://6a3d469f30df9b9597fdcd91442ab9f973a9dbde@github.com/cosmos/voyager.git'

Error: remote: error: GH006: Protected branch update failed for refs/heads/master.        
remote: error: 3 of 3 required status checks are expected.        
To https://6a3d469f30df9b9597fdcd91442ab9f973a9dbde@github.com/cosmos/voyager.git
 ! [remote rejected] HEAD -> master (protected branch hook declined)
error: failed to push some refs to 'https://6a3d469f30df9b9597fdcd91442ab9f973a9dbde@github.com/cosmos/voyager.git'

    at /home/circleci/project/node_modules/simple-git/promise.js:21:26
    at Git.<anonymous> (/home/circleci/project/node_modules/simple-git/src/git.js:944:21)
    at Function.Git.fail (/home/circleci/project/node_modules/simple-git/src/git.js:1452:18)
    at fail (/home/circleci/project/node_modules/simple-git/src/git.js:1410:20)
    at /home/circleci/project/node_modules/simple-git/src/git.js:1419:16
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:182:7)
Exited with code 1
```

**Comment by faboweb** [comment]
How about removing the the status checks on master and limiting who can merge to master to set of people including our bot?

**Comment by faboweb** [comment]
I went ahead and enabled this change.

**Comment by NodeGuy** [comment]
This is good to go.  @faboweb I suggest you reset `master` via:

```
git push -f origin 0.6.2:master
```

**Comment by faboweb** [comment]
I will remove ci files from coverage in the short future.

**Pull request merged**